### PR TITLE
Win32a changes submitted

### PR DESCRIPTION
**Pull request by Bill-Gray**
With this,  "mainstream" PDCurses should be updated to include the new Win32a flavor.  Other flavors are not greatly affected,  except as described at http://projectpluto.com/win32a.htm#2015may31 and in the commit comments.  So I _think_ this will be straightforward,  even though it's a fair bit of code.

### CODE REVIEW

**Comment by TheGentzy** [comment]
I've been using this recently and its been running smoothly. I did have a few issues getting it to work along the way, but I believe those were my mistakes.

**Comment by techtonik** [comment]
This needs to be rebased to solve conflicts.

**Comment by luke-jr** [comment]
Merged, not rebased. Rebasing is considered a bad practice.

File: curses.h [hunk 1]
```diff
@@ -34,11 +34,16 @@ PDCurses portable platform definitions list:
 #define XOPEN           1      /* X/Open Curses routines */
 #define SYSVcurses      1      /* System V Curses routines */
 #define BSDcurses       1      /* BSD Curses routines */
-#define CHTYPE_LONG     1      /* size of chtype; long */
+#if defined( CHTYPE_32)
+   #define CHTYPE_LONG     1      /* chtypes will be 32 bits */
+#elif !defined( CHTYPE_16)
+   #define CHTYPE_LONG     2      /* chtypes will be (default) 64 bits */
```
Comment (luke-jr) [inline curses.h]: What if CHTYPE_16 is defined?

File: curses.h [hunk 2]
```diff
@@ -83,11 +88,15 @@ extern "C"
...
-# else
-typedef unsigned long chtype;  /* 16-bit attr + 16-bit char */
-# endif
+    #if(CHTYPE_LONG >= 2)       /* "non-standard" 64-bit chtypes     */
+        typedef uint64_t chtype;
+    #else                        /* "Standard" CHTYPE_LONG case,  32-bit: */
+        # if _LP64
+            typedef unsigned int chtype;
+        # else
+            typedef unsigned long chtype;  /* 16-bit attr + 16-bit char */
```
Comment (luke-jr) [inline curses.h]: You're including stdint.h, so why not use uint32_t?

**Comment by luke-jr** [comment]
BTW, is it possible to build this such that the console vs GDI decision can be made at runtime (eg, via a command-line option)?

**Comment by techtonik** [comment]
> Merged, not rebased. Rebasing is considered a bad practice.

I'd say that merges and huge commits are hard to review, so rebasing and splitting changes into small independent PRs is often the only way forward. Depends on reviewer, though.

**Comment by techtonik** [comment]
> BTW, is it possible to build this such that the console vs GDI decision can be made at runtime (eg, via a command-line option)?

Technically it is possible, but I am not sure that it is a good fit for this PR.

**Comment by Bill-Gray** [comment]
Hi Luke,

On 2016-03-23 00:31, Luke-Jr wrote:
> BTW, is it possible to build this such that the console vs GDI decision can be made at
> runtime (eg, via a command-line option)?

    I suppose it could be done,  with a great deal of revision.  (A _lot_
of revision.)  There's also not a lot of duplicate code involved,  so
whichever flavor you chose,  you'd be loading in a lot of code for the
other flavor without using it.  I guess I'd have to ask "why would you
want to do this?"

    Note that SDL would also be a choice,  so you could load up all
three sets of code and make a runtime decision amongst them.  (There
is actually apt to be more overlap between SDL,  GDI,  and X11 than
between any of them and Win32 console.  All three flavors will use
similar logic for blinking,  determining background/foreground
colors,  and a few other items.  All three could support user and
programmatic resizing,  under/over/left/right/strikeout lining,
"real" blinking,  and dimmed,  italic,  and bold text.  Win32
console can't support most of these,  and has little code overlap
with those three flavors.)

File: curses.h (hunk 1 above)
Comment (Bill-Gray) [inline curses.h]: Hi Luke,

On 2016-03-23 00:26, Luke-Jr wrote:
> In curses.h <https://github.com/wmcbrine/PDCurses/pull/6#discussion_r57108353>:
>
>> @@ -34,11 +34,16 @@ PDCurses portable platform definitions list:
>>  #define XOPEN           1      /* X/Open Curses routines */
>>  #define SYSVcurses      1      /* System V Curses routines */
>>  #define BSDcurses       1      /* BSD Curses routines */
>> -#define CHTYPE_LONG     1      /* size of chtype; long */
>> +#if defined( CHTYPE_32)
>> +   #define CHTYPE_LONG     1      /* chtypes will be 32 bits */
>> +#elif !defined( CHTYPE_16)
>> +   #define CHTYPE_LONG     2      /* chtypes will be (default) 64 bits */
>
> What if CHTYPE_16 is defined?

    In that case,  CHTYPE_LONG is left undefined.

    There's some discussion of this in curses.h (search for "Originally,
PDCurses used a short" and read from there.)  Until Win32a came along,
your options were to leave CHTYPE_LONG undefined,  in which case you
got 16-bit chtypes and some limited attributes and colors;  or you
could #define CHTYPE_LONG 1 and get 32-bit chtypes,  basic Unicode
support (i.e.,  up to U+FFFF),  and plenty of colors and attributes.

    Since this was a binary decision,  it made sense to have the options
be "either CHTYPE_LONG is defined,  or it isn't."  Making it a ternary
choice,  with 64-bit chtypes as an option,  threw a spanner into the
works.  For backward compatibility,  I just set things up so that
option was enabled with #define CHTYPE_LONG 2.

-- Bill

File: curses.h (hunk 2 above)
Comment (Bill-Gray) [inline curses.h]: Hi Luke,

On 2016-03-23 00:27, Luke-Jr wrote:
> In curses.h <https://github.com/wmcbrine/PDCurses/pull/6#discussion_r57108369>:
>
>> @@ -83,11 +88,15 @@ extern "C"
>>  typedef unsigned char bool;    /* PDCurses Boolean type */
>>
>>  #ifdef CHTYPE_LONG
>> -# if _LP64
>> -typedef unsigned int chtype;
>> -# else
>> -typedef unsigned long chtype;  /* 16-bit attr + 16-bit char */
>> -# endif
>> +    #if(CHTYPE_LONG >= 2)       /* "non-standard" 64-bit chtypes     */
>> +        typedef uint64_t chtype;
>> +    #else                        /* "Standard" CHTYPE_LONG case,  32-bit: */
>> +        # if _LP64
>> +            typedef unsigned int chtype;
>> +        # else
>> +            typedef unsigned long chtype;  /* 16-bit attr + 16-bit char */
>
> You're including stdint.h, so why not use uint32_t?

    Um.  Looks to me as if you have a good point there.  I'd prefer letting
the folks who write 'stdint.h' puzzle out which size is which.  Figuring
out what size various types are isn't as simple as checking _LP64;  there
are different macros to check on different compilers,  and there's a reason
'stdint.h' is often a complicated mess.  It could be we have people out
there thinking they're getting 32-bit chtypes when they're really getting
64-bit chtypes.

    Similarly,  a few lines further down :

typedef unsigned short chtype; /* 8-bit attr + 8-bit char */

    should,  I think,  use uint16_t.  I'll make that change.

    (I didn't do this before,  due to tunnel vision.  I was adding my code
for 64-bit chtypes and didn't look too closely at the existing code
for 32-bit and 16-bit chtypes.)

-- Bill

**Comment by Bill-Gray** [comment]
On 2016-03-23 04:54, anatoly techtonik wrote:
>     Merged, not rebased. Rebasing is considered a bad practice.
>
> I'd say that merges and huge commits are hard to review, so rebasing and splitting
> changes into small independent PRs is often the only way forward. Depends on reviewer,
> though.

    You're both getting beyond my very limited knowledge of git (only
started to use it recently,  though I'm now quite enthusiastic about it).
A search on "git merge vs. rebase" is helping to lessen my ignorance.

    The "Golden Rule of Rebasing",  described here,

https://www.atlassian.com/git/tutorials/merging-vs-rebasing/conceptual-overview

    suggests some possible problems (I think) with rebasing.

    Dunno if this would help,  but... since the time I forked from
wmcbrine/PDCurses/master,  that branch has had three commits.  Two
("warning suppression" and "Parameters not given here",  on 7 Feb)
are now incorporated in Bill-Gray/PDCurses/master.  The remaining
commit,  pulling in Laura Michaels' TTF/Unicode SDL changes,  would
be easy enough for me to add in.

    If I do that,  wmcbrine/PDCurses/master could be rolled back
to where it was when I forked it,  and the merger/rebasing would be
done on the basis of "here are some changes from Bill-Gray/PDCurses/master,
and there has been no activity on wmcbrine/PDCurses/master since this
version was forked."

    Seems to me this ought to simplify the problem,  but as a Git
newbie,  I may be overlooking something.

-- Bill

**Comment by luke-jr** [comment]
The purpose of runtime selection would be to avoid opening a separate window when invoked from a command prompt.

**Comment by techtonik** [comment]
@Bill-Gray I would try to split it into several small logical changes using:
```
git branch smallchange   # create new branch from where you are
git checkout smallchange # switch to new branch
git rebase -i HEAD~50     # edit last 50 commits on new branch, remove extra split
```
If you fork `smallchange` branch from you current master, all removed commits will still be available from your `master.

**Comment by Bill-Gray** [comment]
Hi Anatoly,

    There are several problems here.  Some of them may seem troublesome
to me only due to my lack of Git knowledge,  I hope.  The most basic
problem is uncertainty as to whether these changes are welcome in
wmcbrine/PDCurses in the first place.

    I know that William has some reservations about the proposed changes
in Win32a.  I'm not entirely sure what they are.  It may well be that
we will stay forked,  or that William would like some of the changes
but not others.  I'd prefer to combine our efforts,  but should make
it clear that I'm okay with it if these changes are rejected.  I've
drawn on updates submitted to wmcbrine/PDCurses,  such as Laura Michaels'
improvements for SDL and Mark Hessling's X11 improvements,  and expect
to continue to do so.

    I'm not very interested in splitting things up until I know which,
if any,  of the Win32a modifications might actually make it into the
"official" PDCurses.  With that out of the way :

    The changes I've submitted begin with a huge dose of changes made
from March 2010 to January 2016 that predate the move to GitHub.  These
are only documented on my page at

http://www.projectpluto.com/win32a.htm

    which is nearly useless for GitHub purposes.  Ideally,  I'd have been
using GitHub back then,  and all the various changes tracked on that page
would be much more readily tracked via GitHub.  But I didn't do that.
Those changes appear as a massive blob of a commit on 2016 January 13,
"initial Win32a version added".  It would be hard to break those up,
after the fact,  into a series of commits (lovely though that would be).

    Following the move to GitHub,  though,  the commits _are_ a series of
small logical changes.  (I actually was able to separate out some of the
2010 to 2016 work as smaller commits,  though not as much as one would
ideally wish.)

    Since I submitted this pull request back in January,  by the way,  I've
made several dozen commits to Bill-Gray/PDCurses.  If there is actually
some desire to incorporate my changes into wmcbrine/PDCurses,  I'd
be submitting a new pull request that is somewhat more up to date
than the original one.

-- Bill

**Comment by techtonik** [comment]
Hi @Bill-Gray. This will never be merged unless somebody helps you to split this huge piece of work into separate small pull requests - one feature per request. It is very hard to review big pieces of code when people have only got about 15 minutes of free time on their side project such as PDCurses during the day (or even week). I spent 5 from 15 already by answering. =)

**Comment by Bill-Gray** [comment]
On 2016-10-24 04:32, anatoly techtonik wrote:
> Hi @Bill-Gray <https://github.com/Bill-Gray>. This will never be merged
> unless somebody helps you to split this huge piece of work into separate
> small pull requests - one feature per request. It is very hard to review
> big pieces of code when people have only got about 15 minutes of free
> time on their side project such as PDCurses during the day (or even week).
> I spent 5 from 15 already by answering. =)

    I'm pretty much proceeding under the assumption that it won't get
merged, and that the fork is a permanent one.

    I also don't see this as a big problem.  We will effectively have
"classic" PDCurses at wmcbrine/PDCurses,  which doesn't change much except
for occasional bug fixes.  And we'll have "new and improved" PDCurses at
Bill-Gray/PDCurses,  with the Windows GDI flavor and assorted changes.

    I perhaps should have pushed harder for a merge back in March 2010,
when this fork was in its infancy.  Had I done that,  and/or if I'd been
using Git so that there would be a series of small commits rather than
the massive blob of a commit dropped on William last January,  then a
merge might be more likely.  (Or,  at least,  a merge of some of the
features of this fork.)

    But at this point,  breaking up the changes between March 2010 (when
Win32a started) and January 2016 (when I moved to GitHub) would be a huge
project and might not get pulled in anyway.  (Or only certain bits and
pieces would get pulled,  and we'd have two versions anyway.)

    Unless I hear objections,  I'll withdraw this pull request and the
two forks can go their separate ways in peace.

-- Bill

**Comment by timofonic** [comment]
@techtonik  @Bill-Gray @TheGentzy @luke-jr @rexx-org Any news about this? Better platform support would be welcome, it can make portability easier.

**Comment by wmcbrine** [comment]
Although I've taken many of the other changes here into my tree, I have yet to include the actual "Win32a" port, because I would've had to dumb it down to fit it into the intentionally limited framework of 3.5 (in particular, no 64-bit chtype, nor anything that depends on it), only to build it up again later. But I do plan to include it in a future version (soon), that more radically revamps PDCurses.

**Comment by Bill-Gray** [comment]
Glad to hear this is apt to happen,  and I agree with your strategy of getting other bits,  such as 64-bit chtypes,  in place first (and you've been getting some pieces of the puzzle into place:  bold/italic fonts,  real blinking,  etc.)

As best I can tell,  the main obstacle to merging the Win32a port and associated changes is that it's a significant pain in the anatomy.  Nothing that's rocket surgery,  but it's a _lot_ of changes.  I don't think there are major philosophical differences to be bridged (and I'm quite flexible on minor philosophical differences).

If you see changes I could make in my fork that would simplify the process,  please let me know.  (One that I've started on is applying some of your changes,  where possible/appropriate,  to my fork,  thereby making them at least slightly less "different".  There's quite a bit more that I can do there,  though.)

**Comment by GitMensch** [comment]
@Bill-Gray Can you please give a short update about "back-merging changes of this repo to your fork"? I think it would be good to do so (and provide a new release in the other repo to verify that all the changes work before more approaches are done to merge the two repos).

@wmcbrine Can you please give a short update/details about 64bit chtype addition and other changes you see necessary before merging the Bill-Gray fork?

**Comment by Bill-Gray** [comment]
@GitMensch  : not entirely sure what you mean by this...

I've been pleased to see a lot of interesting things in @wmcbrine's updates in the last few months,  and have gotten around to putting some of his fixes and improvements into my fork.  I was rather impressed that William found a way to get blinking text and 256 colors in the Win32 console.  I'd be grabbing still more of his improvements,  but I've some contract work that has been taking most of my time for a while now.  I'm also just glad to see that "official" PDCurses has some life to it.

Unfortunately,  we've now diverged a lot.  Previously,  I'd kept up with William's changes,  so that my fork was essentially "official" PDCurses plus changes.  A merger now would be painful,  though not impossible,  _if_ William is so inclined.  It would probably start with adding 64-bit chtypes.  Doing that would not really be very difficult.   I think that at that point,  the Win32a code could simply be slurped in.  Since "official" PDCurses doesn't have Win32a at all,  there would be no issue with conflicts with existing code.

After that,  we could put dimmed,  overlined,  blinking,  struck-out,  RGB,  full Unicode,  etc. capabilities into other platforms that can support them (basically X11 and SDL1 and 2).  That would take some more serious effort,  but shouldn't be rocket surgery,  especially since all of it does already work in the forked version.  But again,  I've no idea if William is interested in this,  and this _is_ his repo.

**Comment by GitMensch** [comment]
@Bill-Gray just wanted to drop a "wow" for the last push and the effort to streamline with this upstream repo. One single "look out" comment: this repository does not use `assert.h` any more (as far as I've seen), your changes do, which _may_ (or _may not_) produce some compatibility issues in "exotic environments".

**Comment by Bill-Gray** [comment]
@GitMensch : ? `<assert.h>` is standard C,  going back roughly to K&R.
I did get a batch of Travis CI errors.  For DOS,  WinCon,  X11,  and SDL2,  I essentially replaced all the code with William's,  slightly modified.  However,  that caused me to lose the CMakeLists.txt file from each directory.  I've replaced it;  we'll see if that causes CI to work.
There _are_ still some compatibility issues,  which I'm working on.  The breakthrough moment for me was the realization that I could probably pull in William's code and get the WinGUI and VT code to be compatible.   I expect to adopt his code for OS/2 and SDL1,  much as I've done for DOS,  WinCon,  X11,  and SDL2.
In each case,  I'll be modifying the code to put 64-bit chtypes,  more ACS chars,  underlining/strikeout/dimmed text,  full Unicode including wide characters and combining characters and SMP.
Truthfully,  I have doubts this code will ever get pulled into the main branch.  But William has made some excellent improvements,  and I wanted both to get the benefit of them and ensure that the fork didn't get us separated to the point of total incompatibility.  That's really what I'm aiming for here.

**Comment by rexx-org** [comment]
Great work Bill. I REALLY hope William can merge your changes in. It is a real pain to have to support 2 different PDCurses ports in THE. If the 2 codebases get merged then we can get back to having proper build processes so PDCurses can once again be included in Linux distributions, again so that those of who rely on PDCurses can also distribute our software properly.

**Comment by Bill-Gray** [comment]
I'd be a happier camper if PDCurses was simply included in distros.  But I do see that as a separate issue.  The folks doing that inclusion could just grab this version.
Still working on getting things better aligned between the forks.  Much of it is quite straightforward.  I have to wrap my head a bit around the new copy/paste mechanism... which,  I should note,  I'm quite glad to see;  hadn't thought about 'centralizing' it the way William has done,  but it does get rid of some redundant code.
Just posted code that removes the A_RGB scheme,  and added (currently just for the VT and WinGUI flavors) a [COLORS=16777472 (2^24 + 2^8) scheme.](https://github.com/Bill-Gray/PDCurses/blob/master/common/pdccolor.txt) The linked text should explain how all this ought to work and some future steps (some of which I don't see how we'd avoid anyway.)

File: curses.h [hunk 3]
```diff
@@ -28,11 +28,6 @@ Defined by this header:
 **man-end****************************************************************/
 
 #define PDCURSES        1      /* PDCurses-only routines */
-#if defined( CHTYPE_32)
-   #define CHTYPE_LONG     1      /* chtypes will be 32 bits */
-#elif !defined( CHTYPE_16)
-   #define CHTYPE_LONG     2      /* chtypes will be (default) 64 bits */
-#endif
```
Comment (GitMensch) [inline curses.h]: Is the dropping of CHTYPE_16 (in constrast to support for CHTYPE_64) a change that @wmcbrine is fine to include or are there still environments where this would be reasonable to keep?
... and as I'm out of curses handling currently - what chtypes are used in ncurses?

**Comment by Bill-Gray** [comment]
@wmcbrine  [dropped 16-bit `chtype`s in November 2018](https://github.com/wmcbrine/PDCurses/commit/e1bc364634b633eaccde59585f9bdf99b2cc55c6).  I've followed suit because (a) I'm trying to align my fork at least passably well with this one,  and (b) I don't think there's really a use case for the 16-bit flavor at this point.  (A really memory-constrained platform?  It does save you 4K on an 80x24 DOS system,  but arguing that would really be stretching things.)
`ncurses` has 64-bit `chtype`s and `mmask_t`.  I don't recall when that change was made,  but I think it goes back a long time.  It's enabled `ncurses` to support the full range of Unicode (including SMP,  Unicode past 64K) and at least 32K colors... I think more than that on some platforms,  given the existence of [`init_extended_pair()`](http://man7.org/linux/man-pages/man3/init_extended_pair.3x.html).
A 64-bit `mmask_t` would mean that we could,  for example,  #define `BUTTONn_TRIPLE_CLICKED` and `BUTTONn_MOVED` to have distinct values.  We could support up to ten mouse buttons (six bits per button).  However,  my current plan is to first finish aligning my fork with William's,  and give him a chance to pull anything in which he make take an interest.  (That alignment isn't entirely complete.  For example,  the versioning macros for `PDC_BUILD`,  `PDC_VER_MAJOR/MINOR/DOT`,  etc.,  could mirror one another somewhat better.  We still handle ACS header files differently,  and that's complicated by the fact that I added a lot of ACS characters.   But we're maybe 90% of the way there.)
Incidentally,  we can't assume that 64-bit integers are available,  though I know of only one exception (Digital Mars compiler in DOS).
I should note that even if none of this gets pulled upstream,  it'll still have been a worthwhile effort.  I now can use William's non-forking `x11` flavor.  There are a _lot_ of good improvements from his WinCon and SDLn platforms.  With the underlying code increasingly similar on both platforms,  life should be much easier for those maintaining or using both forks.

**Comment by GitMensch** [comment]
Thank you for the explanations. I actually *guess* the most easy way to pull into upstream is your vt port - 64bit chtypes. I guess that is the only part that "needs to be removed" to make it available and working, correct? If not - can you maybe "prepare" that (maybe by a define so we can quite easily pull the changes in whenever he feels ready to do this [in the meantime others would still have the option to combine this port into upstream])?

**Comment by Bill-Gray** [comment]
I'm assuming this is essentially a fossil pull request,  and am closing it.
